File: src/Boundaries/dirichletALEMeshBoundary.py
# Copyright (C) 2026  Cody Raskin

from PYB11Generator import *
from boundary import *

@PYB11template("dim")
class DirichletALEMeshBoundary(Boundary):
    def pyinit(self, mesh="Mesh::ALEMesh<%(dim)s>*",
               rho="double", v="Lin::Vector<%(dim)s>", u="double"):
        return
    def setFaces(self, ids="std::vector<size_t>"):
        return
    def setAllBoundaryFaces(self):
        return

DirichletALEMeshBoundary2d = PYB11TemplateClass(DirichletALEMeshBoundary,
                              template_parameters = ("2"),
                              cppname = "DirichletALEMeshBoundary<2>",
                              pyname = "DirichletALEMeshBoundary2d",
                              docext = " (2D).")

File: src/Boundaries/dirichletALEMeshBoundary.cc
// Copyright (C) 2026  Cody Raskin

#pragma once
#include "aleMeshBoundary.hh"

// Fixed exterior state (pinned inflow): returns the same (rho, v, u) at
// every governed face regardless of the interior state. Same role as
// DirichletGridBoundary for pinned inflow.
template <int dim>
class DirichletALEMeshBoundary : public ALEMeshBoundary<dim> {
protected:
    double rho_;
    Lin::Vector<dim> v_;
    double u_;

public:
    using Vector = Lin::Vector<dim>;

    DirichletALEMeshBoundary(Mesh::ALEMesh<dim>* mesh, double rho, Vector v, double u)
        : ALEMeshBoundary<dim>(mesh), rho_(rho), v_(v), u_(u) {}

    virtual void
    ghostState(size_t faceIndex,
               double rhoIn, const Vector& vIn, double uIn,
               double& rhoOut, Vector& vOut, double& uOut) const override {
        (void)faceIndex; (void)rhoIn; (void)vIn; (void)uIn;
        rhoOut = rho_;
        vOut = v_;
        uOut = u_;
    }
};
